### DISCUSSION TITLE
Dependency sets

### DISCUSSION THREAD

**Discussion by zkochan** [pnpm/pnpm#3958] [Ideas]
Some devs complain that with the non-hoisted node_modules structure they cannot create dependencies that bundle all their tools together. With npm, they can create a package that has eslint, jest, etc as dependencies and then they add this tooling package to their project and jest, eslint are added to node_modules.

Maybe we could support such case somehow.

Using an inherit feature for instance? Where we would inherit dev deps.

Or using a separate dependencies field, where the subdeps are hoisted. Or using some naming convention for such special dev deps.

**Comment by wre232114** [comment]
That is also a useful feature I am thinking about. in my team we develop infra tools, for example, A, B and C, and B depends on A.

B may want to use the same dependencies(for example, webpack) from A instead of declare its own(which may cause multiple instance error).

I think we can add a field that can export dependencies to the parent, the dependencies will be symlinked to parent node_modules so it can be access in parent. if parent already declared this dependency, the declared one will be used.

**Reply by wre232114** [reply]
a field like 'exportDependencies', I am not sure

**Reply by zkochan** [reply]
Peer dependencies should be used in this case.

**Reply by wre232114** [reply]
i think its different from peer dependencies, because we want the parent to use the dependencies from child. while peer dependencies is child use the dependencies from parent

**Reply by zkochan** [reply]
if there is a package that needs to be a singleton, use it as a peer dependency. This is why peer dependencies exist.

**Reply by wre232114** [reply]
it's different, peer dependencies means install the shared dependencies in the parent, the version is controled by user, but we want to make sure that the user's package uses our dependencies

**Reply by wre232114** [reply]
use peer dependencies and check the dependencies will also work, but if the feature in this thread implemented, will make things more easier

**Reply by wre232114** [reply]
> it's different, peer dependencies means install the shared dependencies in the parent, the version is controled by user, but we want to make sure that the user's package uses our dependencies

and user also will not need to install the dependencies i installed, like tools pack you mentioned@zkochan . multiple instance is one example we meet frequently

**Comment by hubertkuoch** [comment]
Just wondering out of curisosity whether that has been implemented or should we keep on using public-hoist-pattern to hoist all dependencies of such packages? Not really an issue but that feature would be indeed quite convenient especially for encapsulating development tool packages.
